src/Agoda.Analyzers.Test/AgodaCustom/AG0021UnitTests.cs

```diff
@@ -0,0 +1,1302 @@
+						return Task.CompletedTask;
+					}
+
+					public void InvocationMethod() {
+						Action TestMethod = () => {};
+
+						TestMethod();
+					}
+				}
+			";
+
+            await TestForResults(code);
+        }
+
+        [Test]
+        public async Task AG0021_Instance_WhenAsyncIsNotAccessible_ShouldNotShowWarning()
+        {
+            var code = @"
+				using System;
+				using System.Threading;
+				using System.Threading.Tasks;
+
+				class TestClassDeclaration {
+					public void TestMethod() {
+						int instance = 1;
+					}
+
+					protected Task TestMethodAsync() {
+						return Task.CompletedTask;
+					}
+				}
+
+				class TestClassInvocation {
+					public void InvocationMethod() {
+						var instance = new TestClassDeclaration();
+
+						instance.TestMethod();
+					}
+				}
+			";
+
+            await TestForResults(code);
+        }
+
+        [Test]
+        public async Task AG0021_Static_WhenAsyncIsNotAccessible_ShouldNotShowWarning()
+        {
+            var code = @"
+				using System;
+				using System.Threading;
+				using System.Threading.Tasks;
+
+				class TestClassDeclaration {
+					public static void TestMethod() {
+						int instance = 1;
+					}
+
+					protected static Task TestMethodAsync() {
+						return Task.CompletedTask;
+					}
+				}
+
+				class TestClassInvocation {
+					public void InvocationMethod() {
+						TestClassDeclaration.TestMethod();
+					}
+				}
+			";
+
+            await TestForResults(code);
+        }
+
+        [Test]
+        public async Task AG0021_Instance_WhenUsedOnClassField_ShouldShowWarning()
+        {
+            var code = @"
+				using System;
+				using System.Threading;
+				using System.Threading.Tasks;
+
+				class TestClassDeclaration {
+					public void TestMethod() {
+						int instance = 1;
+					}
+
+					public Task TestMethodAsync() {
+						return Task.CompletedTask;
+					}
+				}
+
+				class TestClassInvocation {
+					public TestClassDeclaration Instance = new TestClassDeclaration();
+
+					public void InvocationMethod() {
+						Instance.TestMethod();
+					}
+				}
+			";
+
+            var baseResult =
+                CSharpDiagnostic(AG0021PreferAsyncMethods.DiagnosticId);
+            var expected = new[]
+            {
+                baseResult.WithLocation(20, 7)
+            };
+
+            await TestForResults(code, expected);
+        }
+
+        [Test]
+        public async Task AG0021_Instance_WhenUsedOnDynamic_ShouldNotShowWarning()
+        {
+            var code = @"
+				using System;
+				using System.Threading;
+				using System.Threading.Tasks;
+
+				class TestClassDeclaration {
+					public void TestMethod() {
+						int instance = 1;
+					}
+
+					public Task TestMethodAsync() {
+						return Task.CompletedTask;
+					}
+				}
+
+				class TestClassInvocation {
+					public void InvocationMethod() {
+						dynamic instance = new TestClassDeclaration();
```

Comment:
We disallow dynamics completely in AG0030, so you I don't need to test for this.

Reply:
It is so, but 
- I believe our checks shouldn't interact each other. Otherwise we have to review all our rules on any change.
- User can dissable any rule in any time, so other chackes should work properly.
- The checking is important due to `dynamic` may crash the current tule. So it is a part of the regressions.